### PR TITLE
Feature/unittests

### PR DESCRIPTION
**Pull request by stancdam**

### CODE REVIEW

File: DataReader.xcodeproj/xcuserdata/soinhaze.xcuserdatad/xcschemes/DataReader.xcscheme [hunk 1]
```diff
@@ -0,0 +1,114 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (tomasz-czyzak) [inline DataReader.xcodeproj/xcuserdata/soinhaze.xcuserdatad/xcschemes/DataReader.xcscheme]: any file under xcuserdata/soinhaze.xcuserdatad should not be submitted...  optionally you may create public scheme

File: DataReader/Classes/AppDelegate.swift [hunk 2]
```diff
@@ -8,12 +8,23 @@
...
-@UIApplicationMain
 class AppDelegate: UIResponder, UIApplicationDelegate {
 
     var window: UIWindow?
-
+    
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
+        // Initialize Storyboard
+        
+        self.window = UIWindow(frame: UIScreen.main.bounds)
```
Comment (tomasz-czyzak) [inline DataReader/Classes/AppDelegate.swift]: self not needed

File: DataReader/Classes/AppDelegate.swift [hunk 3]
```diff
@@ -8,12 +8,23 @@
...
+        let storyboard = UIStoryboard(name: "Main", bundle: Bundle.main)
+        
+        // Instantiate Initial View Controller
+        if let viewController = storyboard.instantiateInitialViewController() as? ViewController {
```
Comment (tomasz-czyzak) [inline DataReader/Classes/AppDelegate.swift]: casting to ViewController isn't needed here

File: DataReader/Classes/AppDelegate.swift [hunk 4]
```diff
@@ -8,12 +8,23 @@
...
+
+            // Set Root View Controller
+            self.window?.rootViewController = viewController
```
Comment (tomasz-czyzak) [inline DataReader/Classes/AppDelegate.swift]: self isn't needed

File: DataReader/Classes/AppDelegate.swift [hunk 5]
```diff
@@ -8,12 +8,23 @@
...
+            self.window?.makeKeyAndVisible()
+        }
        return true
```
Comment (tomasz-czyzak) [inline DataReader/Classes/AppDelegate.swift]:
```
        window = UIWindow()
        let storyboard = UIStoryboard(name: "Main", bundle: Bundle.main)
        let viewController = storyboard.instantiateInitialViewController()
        window?.rootViewController = viewController
        window?.makeKeyAndVisible()
        return true
```

File: DataReader/Controller/ViewController.swift [hunk 6]
```diff
@@ -16,20 +16,17 @@ class ViewController: UIViewController {
...
-    
+    var apiService: APIServiceProtocol! = APIService()
+    var coreDataStack: CoreDataStackProtocol! = CoreDataStack()
     
     // MARK: - View related
     
     override func viewDidLoad() {
         super.viewDidLoad()
+        
+        assert(coreDataStack != nil, "dataProvider is not allowed to be nil at this point")
```
Comment (tomasz-czyzak) [inline DataReader/Controller/ViewController.swift]: I'm not a fan of asserts  ;) you should handle it more gentle instead crash

File: DataReader/Controller/ViewController.swift [hunk 7]
```diff
@@ -43,125 +40,42 @@ class ViewController: UIViewController {
...
+            coreDataStack.clearData()
             self.activitiIndicator.isHidden = false
-            APIService.requestData(delegate: self)
+            let jsonUrlString = "https://private-5e934f-datatextapi.apiary-mock.com/data"
+            guard let url = URL(string: jsonUrlString) else { return }
+
+            self.apiService.requestData(url: url) { (data, error) -> Void in
+                guard let data = data else { return }
+                DispatchQueue.main.async {
+                    self.coreDataStack.saveInCoreDataWith(data: data)
```
Comment (tomasz-czyzak) [inline DataReader/Controller/ViewController.swift]: using self here is dangerous... 
consider scenario:what if view controller won't be visible? for example user jumped to another VC

File: DataReader/Controller/ViewController.swift [hunk 8]
```diff
@@ -43,125 +40,42 @@ class ViewController: UIViewController {
...
+        let numberOfObjects = coreDataStack.getNumberOfObjects()
+        if numberOfObjects == 0 {
+            let jsonUrlString = "https://private-5e934f-datatextapi.apiary-mock.com/data"
+            guard let url = URL(string: jsonUrlString) else { return }
             
-            if let results = self.fetchedResultsController.sections {
-                if results[0].numberOfObjects == 0 {
-                    APIService.requestData(delegate: self)
-                } else {
+            self.apiService.requestData(url: url) { (data, error) -> Void in
```
Comment (tomasz-czyzak) [inline DataReader/Controller/ViewController.swift]: duplicate code

File: DataReader/Model/APIService.swift [hunk 9]
```diff
@@ -13,30 +13,44 @@ enum DataManagerError: Error {
...
-        let jsonUrlString = "https://private-5e934f-datatextapi.apiary-mock.com/data"
-        guard let url = URL(string: jsonUrlString) else { return }
+protocol URLSessionProtocol {
+    typealias DataTaskResult = (Data?, URLResponse?, Error?) -> Void
+    
+    func dataTask(with url: URL,
+                  completionHandler: @escaping DataTaskResult) -> URLSessionDataTaskProtocol
+}
+
+extension URLSessionDataTask: URLSessionDataTaskProtocol { }
```
Comment (tomasz-czyzak) [inline DataReader/Model/APIService.swift]: empty extension?

File: DataReader/Model/APIService.swift [hunk 10]
```diff
@@ -13,30 +13,44 @@ enum DataManagerError: Error {
     case invalidJson
 }
 
-protocol FetchManagerDelegate {
-    func requestCompleted(data: [String]?, error: DataManagerError?)
+typealias DataTaskResult = (Data?, URLResponse?, Error?) -> Void
+
+protocol URLSessionDataTaskProtocol {
```
Comment (tomasz-czyzak) [inline DataReader/Model/APIService.swift]: too many protocols ;)

File: DataReader/Model/APIService.swift [hunk 11]
```diff
@@ -13,30 +13,44 @@ enum DataManagerError: Error {
...
+    init(session: URLSessionProtocol = URLSession.shared ) {
+        self.session = session
+    }
+    
+    func requestData(url: URL, callback: @escaping completeClosure)  {
+        self.session.dataTask(with: url) { (data, response, error) in
+            if let _ = error {
+                callback(nil, DataManagerError.unknown)
+            } else {
+                callback(data, nil)
```
Comment (tomasz-czyzak) [inline DataReader/Model/APIService.swift]: what if data is nil? for example response.statusCode = 400 ?
you may test it by changing on statusCode on apiary

File: DataReader/Model/APIService.swift [hunk 12]
```diff
@@ -13,30 +13,44 @@ enum DataManagerError: Error {
...
+            }
+        }.resume()
```
Comment (tomasz-czyzak) [inline DataReader/Model/APIService.swift]: if you pass your own session instead shared then you will have memory leak
details: https://stackoverflow.com/questions/39409357/nsurlsession-http-2-memory-leak/46613760#46613760

File: DataReaderTests/APIServiceTests.swift [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+
+import CoreData
+import XCTest
+@testable import DataReader
+
+class APIServiceTest: XCTestCase {
+        
+    func test_requestData_withResponseData_returnsData() {
+        
+        let session = MockURLSession()
```
Comment (tomasz-czyzak) [inline DataReaderTests/APIServiceTests.swift]: duplicated code in both tests read about setUp and tearDown

File: DataReaderTests/ViewControllerTests.swift [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+    var viewController: ViewController!
+    
+    override func setUp() {
+        super.setUp()
+
+        viewController = UIStoryboard(name: "Main", bundle: nil).instantiateViewController(withIdentifier: "ViewController") as! ViewController
+    }
+    
+    override func tearDown() {
+        super.tearDown()
```
Comment (tomasz-czyzak) [inline DataReaderTests/ViewControllerTests.swift]: empty method?  or maybe missing:
```
viewController = nil
```

File: DataReaderTests/ViewControllerTests.swift [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+    }
+    
+    func test_viewDidLoad_tableViewDataSourceShouldBeSet() {
+
+        let _ = viewController.view
```
Comment (tomasz-czyzak) [inline DataReaderTests/ViewControllerTests.swift]: duplicated line in every test

**Review by tomasz-czyzak — CHANGES_REQUESTED** [review]

File: DataReader/Classes/AppDelegate.swift (hunk 2 above)
Comment (stancdam) [reply DataReader/Classes/AppDelegate.swift]: done

File: DataReader/Classes/AppDelegate.swift (hunk 3 above)
Comment (stancdam) [reply DataReader/Classes/AppDelegate.swift]: done

File: DataReader/Classes/AppDelegate.swift (hunk 4 above)
Comment (stancdam) [reply DataReader/Classes/AppDelegate.swift]: done

File: DataReader/Classes/AppDelegate.swift (hunk 5 above)
Comment (stancdam) [reply DataReader/Classes/AppDelegate.swift]: done

File: DataReader/Controller/ViewController.swift (hunk 7 above)
Comment (stancdam) [reply DataReader/Controller/ViewController.swift]: done

File: DataReader/Controller/ViewController.swift (hunk 8 above)
Comment (stancdam) [reply DataReader/Controller/ViewController.swift]: done

File: DataReader.xcodeproj/xcuserdata/soinhaze.xcuserdatad/xcschemes/DataReader.xcscheme (hunk 1 above)
Comment (stancdam) [reply DataReader.xcodeproj/xcuserdata/soinhaze.xcuserdatad/xcschemes/DataReader.xcscheme]: done

File: DataReaderTests/ViewControllerTests.swift (hunk 15 above)
Comment (stancdam) [reply DataReaderTests/ViewControllerTests.swift]: fixed

File: DataReaderTests/ViewControllerTests.swift (hunk 14 above)
Comment (stancdam) [reply DataReaderTests/ViewControllerTests.swift]: fixed

File: DataReader/Controller/ViewController.swift (hunk 6 above)
Comment (stancdam) [reply DataReader/Controller/ViewController.swift]: hm.. removed assert, done?

File: DataReader/Model/APIService.swift (hunk 9 above)
Comment (stancdam) [reply DataReader/Model/APIService.swift]: explained

File: DataReader/Model/APIService.swift (hunk 10 above)
Comment (stancdam) [reply DataReader/Model/APIService.swift]: hm... cannot agree

File: DataReader/Model/APIService.swift (hunk 11 above)
Comment (stancdam) [reply DataReader/Model/APIService.swift]: fixed

**Pull request merged**